janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java

```diff
@@ -441,43 +439,43 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
     private class CQLPagingIterator implements Iterator<Row> {
 
         private ResultSet currentResultSet;
-
-        private int index;
-        private int paginatedResultSize;
+        private int availableWithoutFetching;
         private final Supplier<Statement> statementSupplier;
 
-        private byte[] lastPagingState = null;
-
-        public CQLPagingIterator(final int pageSize, Supplier<Statement> statementSupplier) {
-            this.index = 0;
-            this.paginatedResultSize = pageSize;
+        public CQLPagingIterator(Supplier<Statement> statementSupplier) {
             this.statementSupplier = statementSupplier;
-            this.currentResultSet = getResultSet();
+            fetchResultSet();
         }
 
         @Override
         public boolean hasNext() {
-            return !currentResultSet.isExhausted();
+            if(availableWithoutFetching<=0){
+                if(currentResultSet.isFullyFetched()){
+                    return false;
+                }
+                fetchResultSet();
```

Comment:
LGTM 👍  I tested locally using the following code which creates and reads 100 thousand vertices. The test passed. I could also see fetchResultSet() is called in hasNext() method.
```java
   @Test
    public void testHugeGraph() {
        int maxK = 100, maxI = 1000;
        for (int k = 0; k < maxK; k++) {
            for (int i = 0; i < maxI; i++) {
                tx.addVertex("prop", k * maxI + i);
                newTx();
            }
        }

        List<Object> props = graph.traversal().V().values("prop").toList();
        assertEquals(maxK * maxI, new HashSet<>(props).size());
    }
```

That being said, I also observed something weird: when I ran this locally, I put a debug point (in Intellij) on this line (line 456) to check if this line is reachable. Once I saw the debug point reached, I immediately removed the debug point and resumed the application. Then the test case failed because `graph.traversal().V().values("prop").toList()` returned fewer results than expected. I could always reproduce it; even when I set `maxI = 10` (that is, one thousand vertices in total). However, the same test case always passes as long as I don't set a debug point. I tried letting the thread sleep for 1 or more seconds here, but the test case still passes. Do you have any thoughts?

Reply:
It's a default behaviour in Intellij debugger to execute the debugged line to print `toString()` result to convince (most likely you hovered your mouth over this line and thus it executed the line).
In you example, you set a debug point on line 456, then you reach that line and when you hover your mouth it will execute `fetchResultSet()` to check if it can print anything meaningful. After that your resume your debug and go fetching a new page (without actually processing any elements on this page). Basically, calling `fetchResultSet()` 2 times instead of 1 time.
That is why it contains fewer elements (i.e. 1 page fewer elements). Moreover, it's actually can produce more elements (in case you are debugging the last page in multi-page traversal).
That said, it's just some debugging problems and not related to the code directly. Also it's just my assumption, in theory it could be something else

Reply:
Got it thanks, that makes sense!